### DISCUSSION TITLE
Kernel modules building in containers running on Colima

### DISCUSSION THREAD

**Discussion by Zherya** [abiosoft/colima#744] [Q&A]
Hi!

I want to build some kernel modules for Colima VM. And for that purpose required files such as Linux headers in `/lib/modules/<kernel version>/build/` are missing. I'm new to Alpine, but as far as I know, they can be installed with `apk` packages `linux-headers` and `linux-virt-dev`.

I figured out that for for the latest Colima VM you use an image of Alpine Linux 3.18.0 with kernel `6.1.29-0-virt`, but used `apk` repositories (https://dl-cdn.alpinelinux.org/alpine/v3.18/main and https://dl-cdn.alpinelinux.org/alpine/v3.18/community) point to packages for Alpine Linux 3.18.2. And `linux-headers` and `linux-virt-dev` packages are now for `6.1.35` kernel. Local packages in`/media/sr0/apks` from OS image also don't contain `linux-headers` and `linux-virt-dev`.

Is it possible to upgrade VM kernel version in order to meet packages version? I've tried some methods but all fail.

Or, maybe, can I somehow get packages for the current kernel version? I was really surprised, but I am unable to find suitable versions of packages, only the latest are available.

Thanks!

**Comment by Zherya** [comment]
If it is not possible to update kernel for already created VM, I wonder if it possible to use another, newer Alpine image: I found [that discussion](https://github.com/abiosoft/colima/discussions/379) that answers my question partially.

And I wonder if I can build [Alpine Lima image](https://github.com/abiosoft/alpine-lima) with, for example, `ALPINE_VERSION=3.18.2` and then use that image for Colima as described in the [discussion](https://github.com/abiosoft/colima/discussions/379) I mentioned.

The only question here (at least, for now ^^), what image edition should be built and what edition is placed in releases in [Alpine Lima](https://github.com/abiosoft/alpine-lima) and then used by Colima? Is it `std`?

**Reply by abiosoft** [reply]
The `colima` branch should be used instead and the edition is `clm`.

**Comment by Zherya** [comment]
Or another option is to use [aports for 3.18.0 release](https://github.com/alpinelinux/aports/tree/8b72c4075123f7310c54ae09e83016890cadefd3) to build required packages manually? I am not sure if it is a simple way...

**Comment by Zherya** [comment]
I have built new image with `Alpine 3.18.2`, used it in `$HOME/.lima/_config/override.yaml` and successfully started Colima with it. However, `linux-virt-dev` package installation was not successful due to read-only filesystem in `/lib/modules/`. I assume that, unfortunately, it is a deadlock for me which cannot be bypassed. Am I right?

**Reply by abiosoft** [reply]
If it is something that can happen at runtime (rather than boot time), a writeable mount can be made for the directory to get the package to install.

**Reply by Zherya** [reply]
Yes, it happens at runtime. I've tried to re-mount it at runtime, but failed due to usage dependency. As far as I remember (I stopped trying to solve the issue some time ago :) ), it was due to `modloop` working and preventing to re-mount. I wonder if I can change `Colima` config or code somehow that it will boot the VM with writable directory.
